### ISSUE TITLE
[QUESTION] How Counter() decides 'consensus' with the same number?

### ISSUE THREAD

**Issue by naliamm** [betteridiot/b575f19#15]
# Instructions
* The more information/context you provide, the better we (the instructors and class) can collectively assist you.

* Please fill out all of the relevant sections below. Incomplete questions will not be addressed by the instruction team. 

* Questions asked ***without*** showing an attempt to answer individually will also not be addressed.

* It is bad etiquette to ask things such as "I don't know how to do this, can you tell me how?" or "I can't get this code to work, what is wrong?" if you do not provide the context and/or screenshots necessary to 
help. 

---
# Question (continued)
Hi! I've noticed that `genome_positions[48]` 'tumor' read has the same number of counts for C (C=21) and A (A=21). 21 is also the highest number in the 'tumor' `Counter()`. I use `.most_common(1)` to get the consensus, so how does `Counter()` decides between A and C? 

```python

```

---
# What do you have?
Use this space to describe what resources/functions/code you have that you feel may be relevant to answering this question. 

Attached below is the output when I call `genome_positions[48]`

**Additionally**, if this is a coding question, describe what **input** this problem section of code requires. By "input", we mean what `type` of data it is.

```python
{'normal': {'consensus': 'T', 'counts': Counter({'T': 3}), 'depth': 3},
 'tumor': {'consensus': 'C',
           'counts': Counter({'C': 21, 'A': 21, 'G': 20, 'T': 17}),
           'depth': 79}}

```

---
# What do you need?
Use this space to describe what resources/functions/code/data you believe is required to determine if this question has been solved.

**Additionally**, if this is a coding question, describe what **output** this problem section of code requires. By "output", we mean what `type` of data it is.

```python
# Paste your code in this code block (if needed)

```

---
# What have you done?
Use this space to describe what approaches you have done in your attempts to solve this question yourself. This is the **most important* part of your question.

I test some counters, and it looks like if there are two bases that occur the same number of times, `Counter()` will give the first base.

## If this is just a general question
Describe what your *hypothesis* of what the answer is here

## If this is a coding question
We require you to post your code, error messages, and output of your code here.

### User-written code
```python
c1 = Counter('aabbcdefg')
print('c1 most common: ', c.most_common(1))
c2 = Counter('bbaacdefg')
print('c2 most common: ', c2.most_common(1))

```

### Observed error messages
```python-traceback
# Paste the observed error messages here

```

### Observed output of code (other than error messages)
```python
c1 most common:  [('a', 2)]
c2 most common: [('b', 2)]

```

---
# Checklist
Replace the space in between the brackets with an X
* [X] Identified what you already have available to address this question
* [ ] Identified what you need to reach an acceptable answer
* [X] Shown what you have done up to this point
    * [X] A hypothesis if this is a general question
    * [ ] Your code, error messages, and output if this is a coding question

**Comment by betteridiot** [comment]
Honestly, it doesn't matter in the grand scheme. The final results require that count of variant be _more_ than half of the total read for tumor. So, if there is a match, no need to inspect further.

Good question though.

**Comment by naliamm** [comment]
Okay thanks! As long as it doesn't matter for hw4.

**Comment by CooperStansbury** [comment]
**tl;dr** `Counter()` returns ties arbitrarily. More below. 

I know this has already been answered. But I was curious and I found this looking through the [collections documentation](https://docs.python.org/2/library/collections.html#collections.Counter) when I had the same question. The method `Counter.most_common(n)` returns arbitrary result when there are ties. 

Below is from the [collections documentation](https://docs.python.org/2/library/collections.html#collections.Counter):

_**most_common([n]):** "Return a list of the n most common elements and their counts from the most common to the least. If n is omitted or None, most_common() returns all elements in the counter. Elements with equal counts are ordered arbitrarily:"_
```
>>> Counter('abracadabra').most_common(3)
[('a', 5), ('r', 2), ('b', 2)]
```

When two items with the same count are compared, the first item to be returned via `_itemgetter(1)` is sorted "alphabetically" (see [this](https://wiki.python.org/moin/HowTo/Sorting/#Key_Functions)), though I imagine there are some hidden factors here. I tested this quickly with the following:

```python
import collections
cnt_1 = collections.Counter()
cnt_1.update(["G","G", "T", "A","A", "C"])
cnt_1.most_common()

cnt_2 = collections.Counter()
cnt_2.update(["G","G", "T", "A","A", "C"])
cnt_2.most_common()
```

Interestingly, `cnt_1` returns "G", not "A". But `cnt_2` returns "A".  

Below is how the `most_common()` method is written (from [collections source code](https://github.com/python/cpython/blob/master/Lib/collections/__init__.py)):

```python
def most_common(self, n=None):
    '''List the n most common elements and their counts from the most
    common to the least.  If n is None, then list all element counts.
    >>> Counter('abracadabra').most_common(3)
    [('a', 5), ('b', 2), ('r', 2)]
    '''
    # Emulate Bag.sortedByCount from Smalltalk
    if n is None:
        return sorted(self.items(), key=_itemgetter(1), reverse=True)
    return _heapq.nlargest(n, self.items(), key=_itemgetter(1))
```

**Comment by betteridiot** [comment]
Yes. Thank you @CooperStansbury I sometimes forget to bring this out of the context of homework and explain more of the nuts & bolts.
